Implement "-tls_version <x>" parameter to select which TLS protocol v…

jeannotlanglois opened the pull request:
…ersion to use (1.0, 1.1 or 1.2)

Commit "55329031" changed the behavior of TLS_init_context()
to always use TLS v1.2.  However when using TLS with non-SIPP
endpoints this can cause issues if they do not support v1.2.

This change allows explicitly specifying which TLS protocol
version to use on the commandline (default is 1.2).

wdoekes commented on src/sipp.cpp:
```diff
@@ -165,6 +165,7 @@ struct sipp_option options_table[] = {
+    {"tls_version", "Set the TLS protocol version to use (1.0, 1.1, 1.2) -- default is 1.2", SIPP_OPTION_FLOAT, &tls_version, 1},
 #else
     {"tls_cert", NULL, SIPP_OPTION_NEED_SSL, NULL, 1},
     {"tls_key", NULL, SIPP_OPTION_NEED_SSL, NULL, 1},
```
You'll want a tls_version in this `#else` block as well.

wdoekes commented on src/sslsocket.cpp:
```diff
@@ -190,15 +186,35 @@ static int sip_tls_load_crls(SSL_CTX* ctx , const char* crlfile)
+        ERROR("Unrecognized TLS version for client context: %1.1f\n", tls_version);
+        sip_trp_ssl_ctx_client = NULL;
+    }
+
```
Duplicate code.

Perhaps refactor the tls_version `if` into a static function?

jeannotlanglois commented:
@wdoekes 

>You'll want a tls_version in this #else block as well.
Oh yes - I had not realized that.  Indeed.  Fixed.

>Duplicate code. Perhaps refactor the tls_version if into a static function?
Yes - agreed.  Refactored into helper static function.

I've re-ran all my testcases (and also validated the 'else' case where no TLS support is built in).  

I've resubmitted my changes as a new commit (I was initially thinking of squashing everything into a single commit to make git history cleaner and facilitate merging but since I've started this pull request I guess this might not be a good idea as it could break things if I deleted the branch to recreate it afterwards).

It appears that one of the Travis CI testcases stalled for a long time - just for one of the builds (not sure why, wondering if this is a bad coincidence since it's only one of the builds).  I do not know enough to troubleshoot (see https://travis-ci.org/SIPp/sipp/jobs/400449783).

Let me know if there's something that still requires changing.

wdoekes commented:
Re-ran the test and it completed just fine :+1:

wdoekes commented:
Thanks!

jeannotlanglois commented:
Thanks! :)
